OpenZFS 9689

behlendorf opened the pull request:
### Motivation and Context

We've wanted a generic range lock implementation for a long time
which can be used outside the ZPL.  Getting this infrastructure in
place now lays the ground work for some in development features.

OpenZFS-issue: https://illumos.org/issues/9689
OpenZFS-commit: https://github.com/openzfs/openzfs/pull/680

### Description

The ZFS range locking code in zfs_rlock.c/h depends on ZPL-specific
data structures, specifically znode_t.  However, it's also used by
the ZVOL code, which uses a "dummy" znode_t to pass to the range
locking code.
    
We should clean this up so that the range locking code is generic
and can be used equally by ZPL and ZVOL, and also can be used by
future consumers that may need to run in userland (libzpool) as
well as the kernel.

### How Has This Been Tested?

Local 12 hour run of `ztest` which didn't uncover any new issues.
Local ZTS test run on CentOS 7.
This is being submitted for further testing by the bots.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

ahrens commented on include/sys/zfs_znode.h:
```diff
@@ -20,7 +20,8 @@
  * Copyright (c) 2005, 2010, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2012, 2015 by Delphix. All rights reserved.
+ * Copyright (c) 2012, 2018 by Delphix. All rights reserved.
+ * Copyright (c) 2014 Integros [integros.com]
```
you probably meant to just change the Delphix copyright (unless you're just trying to reduce future merge conflicts)

ahrens commented on module/zfs/zfs_znode.c:
```diff
@@ -20,7 +20,8 @@
  * Copyright (c) 2005, 2010, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2012, 2014 by Delphix. All rights reserved.
+ * Copyright (c) 2012, 2018 by Delphix. All rights reserved.
+ * Copyright (c) 2014 Integros [integros.com]
```
same here

behlendorf commented on include/sys/zfs_znode.h:
```diff
@@ -213,6 +214,7 @@ typedef struct znode {
 	list_node_t	z_link_node;	/* all znodes in fs link */
 	sa_handle_t	*z_sa_hdl;	/* handle to sa data */
 	boolean_t	z_is_sa;	/* are we native sa? */
+	boolean_t	z_is_zvol;	/* are we used by the zvol */
```
Accidentally, introduced by one of the reverts I'll drop it again.

behlendorf commented:
Refreshed.  I went and squashed the two reverts in to the OpenZFS change.  It was going to be a fair bit of work to sort of the build issues for each those intermediary steps, and in the end I felt it wasn't worthwhile.  This does make OpenZFS 9869 a little harder to review.  If reviewers feel strongly the reverts should be kept separate I can go through and split this up again.

Additionally, I've added a top commit which reintroduces the free list logic to `rangelock_exit()` which was dropped when porting this change.  This is most likely reason for the previous CI failures, but we shall see.

behlendorf commented:
Refreshed and ready for review.   There were only a few additional changes needed.

* Rebased on master
* Preserved the optimized `rangelock_compare` function.
* Several cstyle fixes
* Fixed incorrectly applied zvol hunk which caused the ZTS failures.
